chain/epoch_manager/src/lib.rs

```diff
@@ -1464,73 +1440,151 @@ impl EpochManager {
+            let block_height = *self.get_block_info(block_hash)?.height();
+            assert!(
+                agg_height < block_height,
+                "#{} {} >= #{} {}",
+                agg_hash,
+                agg_height,
+                block_hash,
+                block_height
+            );
+        }
+
+        let epoch_id = self.get_block_info(block_hash)?.epoch_id().clone();
+        let epoch_info = self.get_epoch_info(&epoch_id)?.clone();
+
+        let mut aggregator = EpochInfoAggregator::new(epoch_id.clone(), *block_hash);
+        let mut cur_hash = *block_hash;
+        Ok(Some(loop {
+            #[cfg(test)]
             {
-                // This means that we reached the previous epoch and still hasn't seen
-                // `aggregator.last_block_hash` and therefore implies either a fork has happened
-                // or we are at the start of an epoch. In this case, the new aggregator should
-                // overwrite the old one.
-                overwrite = true;
-                break;
+                self.epoch_info_aggregator_loop_counter += 1;
             }
-            new_aggregator.update(block_info, &epoch_info, prev_height?);
-            cur_hash = *block_info.prev_hash();
-        }
-        aggregator.merge(new_aggregator, overwrite);
 
-        Ok(aggregator)
-    }
+            // To avoid cloning BlockInfo we need to first get reference to the
+            // current block, but then drop it so that we can call
+            // get_block_info for previous block.
+            let block_info = self.get_block_info(&cur_hash)?;
+            let prev_hash = *block_info.prev_hash();
+            let different_epoch = &epoch_id != block_info.epoch_id();
+
+            if different_epoch || prev_hash == CryptoHash::default() {
+                // We’ve reached the beginning of an epoch or a genesis block
+                // without seeing self.epoch_info_aggregator.last_block_hash.
+                // This implies self.epoch_info_aggregator.last_block_hash
+                // belongs to different epoch or we’re on different fork (though
+                // the latter should never happen).  In either case, the
+                // aggregator contains full epoch information.
+                break (aggregator, true);
+            }
 
-    fn save_epoch_info_aggregator(
-        &mut self,
-        store_update: &mut StoreUpdate,
-        aggregator: EpochInfoAggregator,
-        write_to_storage: bool,
-    ) -> Result<(), EpochError> {
-        if write_to_storage {
-            store_update.set_ser(ColEpochInfo, AGGREGATOR_KEY, &aggregator)?;
-        }
-        self.epoch_info_aggregator = Some(aggregator);
-        Ok(())
+            let prev_info = self.get_block_info(&prev_hash)?;
+            let prev_height = *prev_info.height();
+            let prev_epoch = prev_info.epoch_id().clone();
+
+            let block_info = self.get_block_info(&cur_hash)?;
```

Comment:
you already fetched it in 1558

Reply:
`get_block_info` returns a reference (thus `block_info` has a hold on
`self`) so this does not compile:

```rust
            let block_info = self.get_block_info(&cur_hash)?;
            let prev_hash = *block_info.prev_hash();
            let prev_info = self.get_block_info(&prev_hash)?;
            aggregator.update(block_info, &epoch_info, prev_height);
```

That’s why we have to either get `block_info` again or clone it the
first time we fetch it.  The object isn’t small so fetching it again
seems a better option.